nx/lib/nx.ex

```diff
@@ -6658,6 +6658,224 @@ defmodule Nx do
+      ** (ArgumentError) expected a 2-D tensor for ord: :frobenius, got a 1-D tensor
+
+      iex> Nx.norm(Nx.tensor([3, 4]), ord: :nuclear)
+      ** (RuntimeError) nuclear norm not implemented yet
+
+      iex> Nx.norm(Nx.tensor([[0], [1]]), ord: :nuclear)
+      ** (RuntimeError) nuclear norm not implemented yet
+
+      iex> Nx.norm(Nx.tensor([[0], [1]]), ord: -2)
+      ** (RuntimeError) ord: -2 for 2-D tensor not implemented yet
+  """
+
+  def norm(tensor, opts \\ []) when is_list(opts) do
+    %{shape: s} = t = tensor!(tensor)
+    assert_keys!(opts, [:ord, :axes, :keep_axes])
+
+    default_axes_opts = [axes: nil, keep_axes: false, ord: nil]
+    opts = Keyword.merge(default_axes_opts, opts)
+
+    rank = rank(t)
+
+    unless rank == 1 or rank == 2,
+      do: raise(ArgumentError, "expected 1-D or 2-D tensor, got tensor with shape #{inspect(s)}")
+
+    {ord, axes_opts} = Keyword.pop(opts, :ord)
+
+    case ord do
+      nil when rank == 1 -> norm_integer(t, 2, axes_opts)
+      nil when rank == 2 -> norm_frobenius(t, axes_opts)
+      :frobenius -> norm_frobenius(t, axes_opts)
+      :nuclear -> norm_nuclear(t, axes_opts)
+      _ when ord in [:inf, :neg_inf] -> norm_inf(t, ord, axes_opts)
+      _ when is_integer(ord) -> norm_integer(t, ord, axes_opts)
+      _ -> raise ArgumentError, "unknown ord #{inspect(ord)}"
+    end
+  end
+
+  defp norm_frobenius(%{shape: {_}}, _opts),
+    do: raise(ArgumentError, "expected a 2-D tensor for ord: :frobenius, got a 1-D tensor")
+
+  defp norm_frobenius(%{shape: {_, _}} = t, opts), do: norm_integer(t, 2, opts)
+
+  defp norm_nuclear(_t, _opts), do: raise("nuclear norm not implemented yet")
+
+  defp norm_inf(%{shape: shape} = t, ord, axes_opts) when ord in [:inf, :neg_inf] do
+    aggregate_opts = Keyword.merge(axes_opts, axes: [1])
+
+    aggregate_axes = if tuple_size(shape) == 2, do: &sum(&1, aggregate_opts), else: & &1
+    reduce = if ord == :inf, do: &reduce_max/2, else: &reduce_min/2
+
+    t
+    |> Nx.abs()
+    |> aggregate_axes.()
+    |> reduce.(axes_opts)
+  end
+
+  defp norm_integer(%{shape: {_}} = t, 0, opts) do
+    t
+    |> not_equal(0)
+    |> sum(opts)
+  end
+
+  defp norm_integer(%{shape: {_, _}}, 0, _opts) do
+    raise ArgumentError, "expected 1-D tensor for ord: 0, got a 2-D tensor"
+  end
+
+  defp norm_integer(%{shape: {_, _}} = t, ord, opts) when ord in [1, -1] do
+    function =
+      if ord == 1 do
+        &reduce_max/2
+      else
+        &reduce_min/2
+      end
+
+    t
+    |> Nx.abs()
+    |> sum(axes: [0])
+    |> function.(Keyword.take(opts, [:keep_axes]))
+  end
+
+  defp norm_integer(%{shape: {_, _}}, -2, _opts) do
+    raise "ord: -2 for 2-D tensor not implemented yet"
+  end
+
+  defp norm_integer(%{shape: {_, _}}, ord, _opts) when ord not in -1..2 or ord == 0 do
+    raise ArgumentError, "invalid :ord for 2-D tensor, got: #{inspect(ord)}"
+  end
+
+  defp norm_integer(t, ord, opts) when is_integer(ord) do
+    inv_ord = tensor(1 / ord)
```

Comment:
Sorry I lead you astray last time, this will promote to `f64` rather quickly which you want to avoid. It can be fixed on the user end, but in general it's better to make sure we're not doing it unnecessarily if possible. This is what you want instead: 

```suggestion
  defp norm_integer(%{type: type} = t, ord, opts) when is_integer(ord) do
    output_type = Nx.Type.to_floating(type)
    inv_ord = tensor(1 / ord, type: output_type)
```

Reply:
No worries!